lib/active_merchant/billing/gateways/shift4.rb

```diff
@@ -0,0 +1,354 @@
+        post[:apiOptions] = options[:api_options] if options[:api_options].present?
+      end
+
+      def add_credentials(post, options)
+        post[:dateTime] = options[:date_time] || current_date_time
+        post[:credential] = {}
+        post[:credential][:clientGuid] = options[:credential][:client_guid]
+        post[:credential][:authToken] = options[:credential][:auth_token]
+      end
+
+      def add_datetime(post, options)
+        post[:dateTime] = options[:date_time] || current_date_time
+      end
+
+      def add_transaction(post, options)
+        add_purchase_card(post, options)
+        add_override_bus_date(post, options)
+        add_amex(post, options)
+        add_auto(post, options)
+        add_card_on_file(post, options)
+        add_hotel(post, options)
+        add_pin(post, options)
+      end
+
+      def add_device(post, options)
+        if options[:device].present?
+          post[:device] = {}
+          post[:device][:manufacturer] = options[:device][:manufacturer]
+          post[:device][:model] = options[:device][:model]
+          post[:device][:promptPostalCode] = options[:device][:prompt_postal_code]
+          post[:device][:promptSecurityCode] = options[:device][:prompt_security_code]
+          post[:device][:promptStreetNumber] = options[:device][:prompt_street_number]
+          post[:device][:terminalId] = options[:device][:terminal_id]
+        end
+      end
+
+      def add_emv(post, options)
+        if options[:emv].present?
+          post[:emv] = {}
+          post[:emv][:tlvData] = options[:emv][:tlv_data]
+        end
+      end
+
+      def add_p2pe(post, options)
+        if options[:p2pe].present?
+          post[:p2pe] = {}
+          post[:p2pe][:format] = options[:p2pe][:format]
+          post[:p2pe][:ksn] = options[:p2pe][:ksn]
+        end
+      end
+
+      def add_card(post, credit_card, options)
+        post[:card] = {}
+        if credit_card.respond_to?(:number)
+          post[:card][:expirationDate] = "0#{credit_card.month}#{credit_card.year.to_s[2..4]}"
+          post[:card][:number] = credit_card.number
+          post[:card][:entryMode] = options[:entry_mode]
+          post[:card][:present] = options[:present]
+        else
+          post[:card] = {} if post[:card].nil?
+          post[:card][:token] = {}
+          post[:card][:token][:value] = credit_card
+        end
+        if options[:card].present?
+          post[:card][:expirationDate] = options[:card][:expiration_date] if options[:card][:expiration_date]
+          post[:card][:entryMode] = options[:card][:entry_mode]
+          post[:card][:present] = options[:card][:present]
+          post[:card][:securityCode] = {} unless post[:card][:securityCode]
+          post[:card][:securityCode][:indicator] = options[:card][:indicator] if options[:card][:indicator]
+          post[:card][:securityCode][:value] = options[:card][:value] if options[:card][:value]
+          post[:card][:securityCode] = {}
```

Comment:
Looks like this will empty the hash after appending data to it, so if this were to remain after refactor, you'll want to remove this line as well.